### PR TITLE
Float eq/neq assertions are not accurate

### PR DESCRIPTION
**Pull request by Snaipe**
Two adjascent floats can be considered not equal if the choice of epsilon is unwise. This becomes problematic because the choice of epsilon depends on the order of the two floating point values one is comparing; for instance, between 1e10 and 1e11, an epsilon of 1 is acceptable, while it is not for 1 and 0.9.

Instead, we should be comparing the ULPs of the difference between the two values, effectively using the N next adjascent floats as upper and lower bound for equality.
It should also be noted that FLOAT_MAX and INFINITY are within 1 ULP from each other, so these special cases should also be handled.

Since we are comparing arbitrary floating-point values, there is seemingly no silver bullet, as ULPs are effective for nonzero value comparison, while epsilons are.
Great care should be used while implementing a solution that satisfies any scenario.

An article describing the technique can be found [here](https://randomascii.wordpress.com/2012/02/25/comparing-floating-point-numbers-2012-edition/)

### CODE REVIEW

File: src/float/ieee754.c [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    return x < 0 ? -x : x;
+}
+
+#ifdef __GNUC__
+# define likely(x)   __builtin_expect((x),1)
+# define unlikely(x) __builtin_expect((x),0)
+#else
+# define likely(x)
+# define unlikely(x)
+#endif
```
Comment (am11) [inline src/float/ieee754.c]: Changing this `ifdef` block to:

```C
#ifndef __GNUC__
# define __builtin_expect(exp, c) (exp)
#endif
# define likely(x)   __builtin_expect((x),1)
# define unlikely(x) __builtin_expect((x),0)
```

builds on VS2015.

**Comment by Snaipe** [comment]
This will be on hold until after the next release. Implementing the comparison functions is non-trivial, and I need to research good defaults for each strategy before hoping to merge anything back to bleeding.

Backward compatibility also needs to be addressed for the ``cr_assert_float_*`` assertion family.

File: test/float.cc [hunk 2]
```diff
@@ -0,0 +1,69 @@
+#include "criterion/parameterized.h"
+#include "criterion/options.h"
+#include "criterion/theories.h"
+#include "criterion/float.h"
+#include <cmath>
+#include <cfloat>
+#include <iostream>
+#include <unistd.h>
```
Comment (am11) [inline test/float.cc]: This can probably be changed to:

```c
#include <iostream>

#ifdef _MSC_VER
# include <io.h>
#else
# include <unistd.h>
#endif
```

Comment (Snaipe) [inline test/float.cc]: Actually unistd.h isn't used at all, it's just a forgotten artifact that I'll most probably remove later.

**Comment by Snaipe** [comment]
Closing this since #182 provides all of what this PR promises, but better.

**Pull request closed without merging**